Swift 5.9 release blog post (#372)
apple/swift-org-website: _posts/2023-08-29-swift-5.9-released.md

---
layout: post
published: true
date: 2023-08-29 15:00:00
title: Swift 5.9 Released!
author: [alexandersandberg]
---

Swift 5.9 is now officially released! 🎉 <!-- TODO: Summary of what's included in the release -->

Thank you to everyone in the Swift community who made this release possible. Your Swift Forums discussions, bug reports, pull requests, educational content, and other contributions are always appreciated!

<!-- TODO: Link to Swift 5.9 showcase resource -->

[The Swift Programming Language](https://docs.swift.org/swift-book/documentation/the-swift-programming-language/) book has been updated for Swift 5.9 and is now published with DocC. This is the official Swift guide and a great entry point for those new to Swift. The Swift community also maintains a number of [translations](/documentation/#translations).

If you’re new to Swift, [The Swift Programming Language](https://docs.swift.org/swift-book/) is the official Swift guide and has been updated for version 5.9. The Swift community maintains a number of [translations](https://www.swift.org/documentation/#translations).

## Language and Standard Library

> <!-- TODO: Add section content -->

Reviewer: Is it worth calling out some of the concurrency changes? I don't know how deep we want to go in these sections though.

```suggestion
<!-- TODO: Add section content -->

Actors can now declare a custom executor which allow for fine grained control over where–e.g. a specific thread or serial dispatch queue–tasks running on an actor will be executed.
```

maybe? The other concurrency changes are somewhat more minor so it's okey to not call them out explicitly I think.

and maybe some notes about how we've kept improving diagnostics around concurrency and sendable etc? @hborla would know if something worth calling out explicitly -- maybe good to say some "progress on sendable diagnostics towards the Swift 6 model"? 🤔

Author: Thanks for the ideas, @ktoso! @beccadax is working on the content for this section, so maybe she can include this. 🙂

Reviewer: Cool, works for me! Thanks @beccadax !

Author: Now added in https://github.com/apple/swift-org-website/pull/372/commits/e1bc48b47c2ea31b5fa15ff5d1e0ebedc34c6b97

Feel free to review, @ktoso. 🙂

Reviewer: LGTM, thank you! :)

---
layout: post
published: true
date: 2023-08-29 15:00:00
title: Swift 5.9 Released!
author: [alexandersandberg]
---

Swift 5.9 is now officially released! 🎉 <!-- TODO: Summary of what's included in the release -->

Thank you to everyone in the Swift community who made this release possible. Your Swift Forums discussions, bug reports, pull requests, educational content, and other contributions are always appreciated!

<!-- TODO: Link to Swift 5.9 showcase resource -->

[The Swift Programming Language](https://docs.swift.org/swift-book/documentation/the-swift-programming-language/) book has been updated for Swift 5.9 and is now published with DocC. This is the official Swift guide and a great entry point for those new to Swift. The Swift community also maintains a number of [translations](/documentation/#translations).

If you’re new to Swift, [The Swift Programming Language](https://docs.swift.org/swift-book/) is the official Swift guide and has been updated for version 5.9. The Swift community maintains a number of [translations](https://www.swift.org/documentation/#translations).

## Language and Standard Library

> <!-- TODO: Add section content -->

Reviewer 2: Worth adding something about on-crash backtraces here also. For Swift 5.9 they're enabled by default on Linux and can be explicitly enabled on macOS for locally built code (though you also need to build using Xcode, or manually add an entitlement). I'll ping @beccadax about it.

Reviewer 3: Is there something I can look at/link to for details on this?

Reviewer 2: I think maybe I should write a separate blog post about the feature and then we can just link to that.

Author: @al45tair It is highly unlikely that we will be able to publish that blog post before this Swift 5.9 one, but it would still be really nice to have it posted afterward. We can chat more about this in the forum DMs.

Considering this, it would be great to have a short summary of the coming blog post somewhere in this blog post, and then also mention that another blog post is coming. This is [what we're doing](https://github.com/apple/swift-org-website/pull/372/files#diff-7b5cf397331a2c12db465ddba80ad03c867a968d86930447676740df1e8a4171R85) for the coming debugging post.

## Language and Standard Library

<!-- TODO: Add section content -->

You can find the complete list of Swift Evolution proposals that were implemented in Swift 5.9 in the [Swift Evolution Appendix](#swift-evolution-appendix) below.

## Developer Experience

<!-- TODO: Add more sections with content? -->

### Debugging

In Swift 5.9, we introduced a couple of new features to [LLDB](https://lldb.llvm.org/) and the Swift compiler aimed at making Swift debugging faster and more reliable.

The `p` and `po` commands will now bypass the Swift compiler when evaluating simple expressions, which will make printing local and member variables as fast as using the `frame variable` or `v` command.

Swift expressions can now refer to generic type parameters. This allows setting a conditional breakpoint in a generic function that only triggers when a type parameter is instantiated with a specific concrete type.

Finally, the debug info produced by the Swift compiler is now more precise when it comes to scoping of local variables.

> For a more detailed discussion and examples of how to use these new features we will be publishing a separate blog focussing on debugging Swift 5.9 next week!

Reviewer 4: @cthielen You might want to weigh in about the exact wording used here ("next week").

Reviewer 5: features we -> features, we
blog focussing -> blog post focusing

```swift
/// Takes any number of `Optional` arguments and returns `nil` if any of them
/// are `nil`, or a tuple of their values if all of them are non-`nil`.
func all<each Wrapped>(_ optional: repeat (each Wrapped)?) -> (repeat each Wrapped)? {
  /// Accumulates the results from the `checkNilness(of:)` helper function.
  var anyAreNil = false

  // Check if any of the arguments in `each optional` is nil by running this
  // `checkNilness(of:)` function on each argument.
  func checkNilness<OneWrapped>(of oneOptional: OneWrapped?) {
      if oneOptional == nil {
          anyAreNil = true
      }
  }
  repeat checkNilness(of: each optional)

  // Did `checkNilness(of:)` find any nils? If so, return nil early.
  if anyAreNil {
      return nil
  }

  // Otherwise, we can safely force-unwrap all of the arguments and form a tuple
  // from the unwrapped values.

>   return (repeat (each optional)!)

Reviewer 3: I chose this example because it’s less of a toy than most of what’s in the proposals, but the downside is that the implementation is relatively complex. Do we have a good way to hide the implementation by default but still have it visible on demand? For instance, maybe in this example we just have:

```swift
func all<each Wrapped>(_ optional: repeat (each Wrapped)?) -> (repeat each Wrapped)? {
    // Implementation provided below
}
```

But then there’s a second code listing with the full implementation behind a `<details>` tag?

Author: I see! Something like this?

https://github.com/apple/swift-org-website/assets/35671299/06108d47-12f9-4fab-b102-86792acdc109

Reviewer 3: Perfect, thanks!

Author: Done in https://github.com/apple/swift-org-website/pull/372/commits/dd2d3a187446ede8c179942efdd87ffbcdaa61a0. Feel free to review. 🙂

}

let optionalInt: Int? = 1
let optionalDouble: Double? = 2.5
let optionalString: String? = nil

if let (int, double, string) = all(optionalInt, optionalDouble, optionalString) {
    print(int, double, string)
}
else {
    print("got a nil")
}
```

What's more, these features are all usable today in Swift 5.9, but they are not finished growing. Each one is too large to fully implement in a single release; we're excited about what you can already do with them in Swift 5.9, but we're even more excited to have laid the foundations for better things to come.

In addition to these major features, Swift 5.9 also includes lots of smaller refinements to the language. The one developers will probably use most often is [`if` and `switch` expressions](https://github.com/apple/swift-evolution/blob/main/proposals/0380-if-switch-expressions.md), which lets you assign the result of a single-expression `if` or `switch` statement to a variable:

```swift
statusBar.text = if !hasConnection { "Disconnected" }
                 else if let error = lastError { error.localizedDescription }
                 else { "Ready" }
```

> Swift 5.9 also adds a [new `package` access level](https://github.com/apple/swift-evolution/blob/main/proposals/0386-package-access-modifier.md) that lets other modules in the same package access APIs, but hides them from code outside the package; it's great for splitting up large modules into several smaller ones without exposing internals to the package's clients. Developers using Swift Concurrency may appreciate the [more convenient `DiscardingTaskGroup` types](https://github.com/apple/swift-evolution/blob/main/proposals/0381-task-group-discard-results.md) for task groups that don't generate results and the advanced [custom actor executors](https://github.com/apple/swift-evolution/blob/main/proposals/0392-custom-actor-executors.md) feature for controlling the exact environment that an actor is run in.

Reviewer 6: I see that `package` access level proposal E-0386 is marked as "Accepted", but not "Implemented in Swift 5.9" yet. Is the implementation incomplete or is there some other reason its implementation status wasn't updated?

Author: @hborla told me that it's fully implemented in Swift 5.9. I guess the status simply isn't updated yet.

Reviewer 7: SE-0386 is implemented in Swift 5.9; I've asked @rjmccall to update the proposal status since he was the review manager.

Reviewer 8: I've updated SE-0386

## Developer Experience

<!-- TODO: Add more sections with content? -->

### Debugging

In Swift 5.9, we introduced a couple of new features to [LLDB](https://lldb.llvm.org/) and the Swift compiler aimed at making Swift debugging faster and more reliable.

The `p` and `po` commands will now bypass the Swift compiler when evaluating simple expressions, which will make printing local and member variables as fast as using the `frame variable` or `v` command.

Swift expressions can now refer to generic type parameters. This allows setting a conditional breakpoint in a generic function that only triggers when a type parameter is instantiated with a specific concrete type.

Finally, the debug info produced by the Swift compiler is now more precise when it comes to scoping of local variables.

For a more detailed discussion and examples of how to use these new features we will be publishing a separate blog focussing on debugging Swift 5.9 next week!

## Ecosystem

<!-- TODO: Add more sections with content? -->

### Swift Package Manager

Following are some highlights from the changes introduced to the [Swift Package Manager](https://github.com/apple/swift-package-manager) in Swift 5.9:

> - SwiftPM packages can now use `package` as a new access modifier, allowing accessing symbols in another target / module within the same package without making it public.

Reviewer 6: This is already mentioned in the "Language and Standard Library" section, is it duplicated on purpose here?

Author: Good callout, we might not need to mention this twice. @tomerd what do you think?

Reviewer 3: FWIW, both the language and the package manager changed to support this feature. At minimum, it might be a good idea to point out the SwiftPM support for `package` in the package manager section.

Reviewer 9: yes, I also think this should stay

Reviewer 5: This should be rewritten to say how SwiftPM was modified to support this new compiler feature, else it is redundant as Max says.

Reviewer 9: [previously](https://github.com/apple/swift-org-website/pull/372/commits/019cfc382d19208ea1568c0398b7873c47f226d7)

---
layout: post
published: true
date: 2023-08-29 15:00:00
title: Swift 5.9 Released!
author: [alexandersandberg]
---

Swift 5.9 is now officially released! 🎉 <!-- TODO: Summary of what's included in the release -->

Thank you to everyone in the Swift community who made this release possible. Your Swift Forums discussions, bug reports, pull requests, educational content, and other contributions are always appreciated!

<!-- TODO: Link to Swift 5.9 showcase resource -->

[The Swift Programming Language](https://docs.swift.org/swift-book/documentation/the-swift-programming-language/) book has been updated for Swift 5.9 and is now published with DocC. This is the official Swift guide and a great entry point for those new to Swift. The Swift community also maintains a number of [translations](/documentation/#translations).

> If you’re new to Swift, [The Swift Programming Language](https://docs.swift.org/swift-book/) is the official Swift guide and has been updated for version 5.9. The Swift community maintains a number of [translations](https://www.swift.org/documentation/#translations).

Reviewer 10: This duplicates the previous paragraph.

Author: Good catch, I think we should remove the first one mentioning DocC as it was already announced in the 5.8 blog post.

Reviewer 5: The second is entirely redundant, I think it should be removed.

/// are `nil`, or a tuple of their values if all of them are non-`nil`.
func all<each Wrapped>(_ optional: repeat (each Wrapped)?) -> (repeat each Wrapped)? {
  /// Accumulates the results from the `checkNilness(of:)` helper function.
  var anyAreNil = false

  // Check if any of the arguments in `each optional` is nil by running this
  // `checkNilness(of:)` function on each argument.
  func checkNilness<OneWrapped>(of oneOptional: OneWrapped?) {
      if oneOptional == nil {
          anyAreNil = true
      }
  }
  repeat checkNilness(of: each optional)

  // Did `checkNilness(of:)` find any nils? If so, return nil early.
  if anyAreNil {
      return nil
  }

  // Otherwise, we can safely force-unwrap all of the arguments and form a tuple
  // from the unwrapped values.
  return (repeat (each optional)!)
}
```

</details>

What's more, these features are all usable today in Swift 5.9, but they are not finished growing. Each one is too large to fully implement in a single release; we're excited about what you can already do with them in Swift 5.9, but we're even more excited to have laid the foundations for better things to come.

> In addition to these major features, Swift 5.9 also includes lots of smaller refinements to the language. The one developers will probably use most often is [`if` and `switch` expressions](https://github.com/apple/swift-evolution/blob/main/proposals/0380-if-switch-expressions.md), which lets you assign the result of a single-expression `if` or `switch` statement to a variable:

Reviewer 11: ```suggestion
In addition to these major features, Swift 5.9 also includes smaller refinements to the language. The one developers will probably use most often is [`if` and `switch` expressions](https://github.com/apple/swift-evolution/blob/main/proposals/0380-if-switch-expressions.md), which lets you assign the result of a single-expression `if` or `switch` statement to a variable:
```

I think "_lots of_" is a matter of subjective opinion (and not necessarily all gravy) and not required here; it's fair to say simply that Swift 5.9 "also includes smaller refinements." Otherwise, I think it reads well.

### C++ Interoperability

Swift 5.9 supports bidirectional interoperability with C++ and Objective-C++. You can now use a subset of C++ APIs in Swift and Swift APIs from C++. For information on enabling C++ interoperability and the supported language subset, please refer to the [documentation](/documentation/cxx-interop/).

C++ interoperability is an actively evolving feature of Swift and developed by the focused [C++ Interoperability Workgroup](/cxx-interop-workgroup/). Certain aspects of the design and functionality might change in future releases of Swift, as the Swift community gathers feedback from real world adoption in mixed Swift and C++ codebases.

Join the Swift community and the C++ Interoperability Workgroup to provide feedback and help shape the evolution of the feature.

> 

Reviewer 9: ```suggestion

### Server

The server workgroup has been busy driving the adoption of Swift concurrency in the server ecosystem, including making use of new 5.9 features such as [custom actor executors](https://github.com/apple/swift-evolution/blob/main/proposals/0392-custom-actor-executors.md). The group recently published their [annual update for 2023](https://www.swift.org/blog/sswg-update-2023/) which goes into details of this work.

```

title: Swift 5.9 Released!
author: [alexandersandberg]
---

Swift 5.9 is now officially released! 🎉 <!-- TODO: Summary of what's included in the release -->

Thank you to everyone in the Swift community who made this release possible. Your Swift Forums discussions, bug reports, pull requests, educational content, and other contributions are always appreciated!

<!-- TODO: Link to Swift 5.9 showcase resource -->

[The Swift Programming Language](https://docs.swift.org/swift-book/documentation/the-swift-programming-language/) book has been updated for Swift 5.9 and is now published with DocC. This is the official Swift guide and a great entry point for those new to Swift. The Swift community also maintains a number of [translations](/documentation/#translations).

If you’re new to Swift, [The Swift Programming Language](https://docs.swift.org/swift-book/) is the official Swift guide and has been updated for version 5.9. The Swift community maintains a number of [translations](https://www.swift.org/documentation/#translations).

## Language and Standard Library

Swift 5.9 tackles three long-desired features that allow you to express things you couldn't easily do before.

> Swift now supports [macros](https://github.com/apple/swift-evolution/blob/main/visions/macros.md), which allow developers to reduce repetitive boilerplate and create new language features that can be distributed as a Swift package. Using a macro is easy and natural; macros can either be called with the function-like freestanding `#macroName` syntax or attached to declarations with an `@MacroName` attribute, so they work just like built-in language features but can't be mistaken for normal code. Implementing a macro uses a powerful and flexible approach: Macros are implemented by writing Swift functions that use the SwiftSyntax library to generate code to insert into the source file. Macros make it easy for your library's users to adopt powerful capabilities that adapt to the code they're used in, like [the new `Observation` module](https://github.com/apple/swift-evolution/blob/main/proposals/0395-observability.md), which allows Swift classes to automatically notify other code when their properties change.

Reviewer 5: natural; -> natural:

Reviewer 5: `A colon (:) is used to introduce information set up by the previous clause. It’s typically used before a list, example, or explanation. `

That's why I suggested it, as [explained here](https://www.grammarly.com/blog/semicolon-vs-colon-vs-dash/).

Reviewer 12: Also from the same page you reference:

> In a sentence, a colon’s job is to introduce information and give it context. Think of it as shorthand for “which is/are,” “as follows,” or “thus.” 

This case doesn’t really work with the above description. They are two independent clauses that could be separate sentences, but the semicolon relates the two more strongly.

> As we mentioned above, a semicolon joins two independent clauses together to create a detailed sentence. Take a look at this example:
> 
> My favorite meal is rigatoni with bolognese sauce and crusty bread; these foods remind me of my grandmother’s cooking. 
> 
> You’ve got two distinct sentences there, but when they’re joined by a semicolon, they paint a fuller picture than they would if kept separate. The semicolon communicates a deeper relationship. Think of it as shorthand for a[ conjunction](https://www.grammarly.com/blog/conjunctions/). Another way to write our example sentence is: My favorite meal is rigatoni with bolognese sauce and crusty bread **because** these foods remind me of my grandmother’s cooking.

Author: Leaving the `;` as is for now, but feel free to continue the discussion.

Reviewer 12: The rule of thumb on that reference page is to use a colon if substituting `“which is/are,” “as follows,” or “thus.”` best replaces the punctuation, and to use a semicolon if “because” best replaces the punctuation.

`Using a macro is easy and natural because macros can either be called with the function-like...`

`Using a macro is easy and natural thus macros can either be called with the function-like...`

So, I agree the second clause explains the reason, the “because", of why macros are easy and natural. But the “because” case uses the semicolon. And yes - grammar points like this can be very subjective!

This feature is also available on macOS, although there it is disabled by default (it can be enabled by setting `SWIFT_BACKTRACE=enable=yes`) and you need to ensure that your program is signed with the `com.apple.security.get-task-allow` entitlement.

### Debugging

In Swift 5.9, we introduced a couple of new features to [LLDB](https://lldb.llvm.org/) and the Swift compiler aimed at making Swift debugging faster and more reliable.

The `p` and `po` commands will now bypass the Swift compiler when evaluating simple expressions, which will make printing local and member variables as fast as using the `frame variable` or `v` command.

Swift expressions can now refer to generic type parameters. This allows setting a conditional breakpoint in a generic function that only triggers when a type parameter is instantiated with a specific concrete type.

Finally, the debug info produced by the Swift compiler is now more precise when it comes to scoping of local variables.

For a more detailed discussion and examples of how to use these new features we will be publishing a separate blog focussing on debugging Swift 5.9 next week!

## Ecosystem

### Swift Package Manager

> Following are some highlights from the changes introduced to the [Swift Package Manager](https://github.com/apple/swift-package-manager) in Swift 5.9:

Reviewer 5: Following -> The following

- SwiftPM packages can now use `package` as a new access modifier, allowing accessing symbols in another target / module within the same package without making it public.

- Introducing `CompilerPluginSupport` module for defining macro targets. Macro targets allow authoring and distribution of custom Swift macros such as [expression macros](https://github.com/apple/swift-evolution/blob/main/proposals/0382-expression-macros.md).

- Introducing new `.embedInCode` resource rule for embedding the contents of the resource into the executable by generating a byte array, e.g.

  ```
  struct PackageResources {
    static let my_resource: [UInt8] = [104,101,108,108,111,32,119,111,114,108,100,10]
  }
  ```

- `allowNetworkConnections(scope:reason:)` setting giving a command plugin permissions to access the network. Permissions can be scoped to Unix domain sockets in general or specifically for Docker, as well as local or remote IP connections which can be limited by port. For non-interactive use cases, there is also a `--allow-network-connections` commandline flag to allow network connections for a particular scope.

> - SwiftPM can now publish to a registry following the publishing spec as defined in [SE-0391](https://github.com/apple/swift-evolution/blob/main/proposals/0391-package-registry-publish.md). SwiftPM also gains support for signed packages. Trust-on-first-use (TOFU) check which includes only fingerprints (e.g., checksums) previously has been extended to include signing identities, and it is enforced for source archives as well as package manifests.

Reviewer 5: Trust-on-first-use -> The Trust-On-First-Use
check which includes only -> check, which only included
previously has -> previously, has

Reviewer 9: https://github.com/apple/swift-org-website/pull/372/commits/a183a95bd46bed851c2947221e2d383b63cdfe23

Reviewer 5: The commas for the "which" clause would make it read better, plus now it should be
has -> have

- New `swift experimental-sdk` experimental command is now available for managing Swift SDK bundles that follow the format described in [SE-0387](https://github.com/apple/swift-evolution/blob/main/proposals/0387-cross-compilation-destinations.md).

See the [Swift Package Manager changelog](https://github.com/apple/swift-package-manager/blob/main/CHANGELOG.md#swift-59) for the complete list of changes.

### Swift Syntax

> [swift-syntax](https://github.com/apple/swift-syntax) has become an essential tool to create macros in Swift 5.9. In addition to the introduction of the modules that allow the creation of macros, swift-syntax has received huge focus on quality:

Reviewer 5: received huge -> received a huge

> - The documentation of swift-syntax has been greatly expanded and can be viewed at [swiftpackageindex.com](https://swiftpackageindex.com/apple/swift-syntax)

Reviewer 5: swift-syntax) -> swift-syntax).

Thank you to everyone in the Swift community who made this release possible. Your Swift Forums discussions, bug reports, pull requests, educational content, and other contributions are always appreciated!

<!-- TODO: Link to Swift 5.9 showcase resource -->

[The Swift Programming Language](https://docs.swift.org/swift-book/documentation/the-swift-programming-language/) book has been updated for Swift 5.9 and is now published with DocC. This is the official Swift guide and a great entry point for those new to Swift. The Swift community also maintains a number of [translations](/documentation/#translations).

## Language and Standard Library

Swift 5.9 tackles three long-desired features that allow you to express things you couldn't easily do before.

> Swift now supports [macros](https://github.com/apple/swift-evolution/blob/main/visions/macros.md), which allow developers to reduce repetitive boilerplate and create new language features that can be distributed as a Swift package. Using a macro is easy and natural; macros can either be called with the function-like freestanding `#macroName` syntax or attached to declarations with an `@MacroName` attribute, so they work just like built-in language features but can't be mistaken for normal code. Implementing a macro uses a powerful and flexible approach: Macros are implemented by writing Swift functions that use the SwiftSyntax library to generate code to insert into the source file. Macros make it easy for your library's users to adopt powerful capabilities that adapt to the code they're used in, like [the new `Observation` module](https://github.com/apple/swift-evolution/blob/main/proposals/0395-observability.md), which allows Swift classes to automatically notify other code when their properties change.

Reviewer 8: ```suggestion
Swift now supports [macros](https://github.com/apple/swift-evolution/blob/main/visions/macros.md), which allow developers to reduce repetitive boilerplate and create new language features that can be distributed as a Swift package. Using a macro is easy and natural; macros can either be called with the function-like freestanding `#macroName` syntax or attached to declarations with an `@MacroName` attribute, so they work just like built-in language features but can't be mistaken for normal code. Implementing a macro uses a powerful and flexible approach: macros are implemented by writing Swift functions that use the SwiftSyntax library to generate code to insert into the source file. Macros make it easy for your library's users to adopt powerful capabilities that adapt to the code they're used in, like [the new `Observation` module](https://github.com/apple/swift-evolution/blob/main/proposals/0395-observability.md), which allows Swift classes to automatically notify other code when their properties change.
```

Reviewer 3: Different style guides disagree on whether you should have a capital letter after a colon, but Apple’s says you should if what’s after the colon forms a complete sentence.

Reviewer 8: That is bizarre to me — I've never seen that outside of very narrow contexts like sentence fragments or titles.  But if it's the official style, it's the style.

Reviewer 2: That's an Americanism, and even then it doesn't seem to be a hard and fast rule. You'd *never* do that in "British" or "International" English — words after colons only get capitalised if there's some other reason for that to happen (e.g. they're a proper noun) — so in some senses it would be better if this wasn't in the Apple style guide because then it'd be correct in both of the major variants of the language. (Likewise, I'm happy with people using -ize instead of -ise in word endings, since that's always been permissible in British usage even though -ise is more common.)

      }
  }
  repeat checkNilness(of: each optional)

  // Did `checkNilness(of:)` find any nils? If so, return nil early.
  if anyAreNil {
      return nil
  }

  // Otherwise, we can safely force-unwrap all of the arguments and form a tuple
  // from the unwrapped values.
  return (repeat (each optional)!)
}
```

</details>

What's more, these features are all usable today in Swift 5.9, but they are not finished growing. Each one is too large to fully implement in a single release; we're excited about what you can already do with them in Swift 5.9, but we're even more excited to have laid the foundations for better things to come.

In addition to these major features, Swift 5.9 also includes smaller refinements to the language. The one developers will probably use most often is [`if` and `switch` expressions](https://github.com/apple/swift-evolution/blob/main/proposals/0380-if-switch-expressions.md), which lets you assign the result of a single-expression `if` or `switch` statement to a variable:

```swift
statusBar.text = if !hasConnection { "Disconnected" }
                 else if let error = lastError { error.localizedDescription }
                 else { "Ready" }
```

> Swift 5.9 also adds a [new `package` access level](https://github.com/apple/swift-evolution/blob/main/proposals/0386-package-access-modifier.md) that lets other modules in the same package access APIs, but hides them from code outside the package; it's great for splitting up large modules into several smaller ones without exposing internals to the package's clients. Developers using Swift Concurrency may appreciate the [more convenient `DiscardingTaskGroup` types](https://github.com/apple/swift-evolution/blob/main/proposals/0381-task-group-discard-results.md) for task groups that don't generate results and the advanced [custom actor executors](https://github.com/apple/swift-evolution/blob/main/proposals/0392-custom-actor-executors.md) feature for controlling the exact environment that an actor is run in.

Reviewer 8: It's a little weird to have a paragraph with just two different things in it.  It doesn't read like a list of unrelated things, but it also doesn't read like a single coherent paragraph.  Maybe split it into two paragraph?

Reviewer 12: I agree that this should be two paragraphs with a new paragraph starting at ‘Developers using Swift Concurrency”.

Reviewer 12: "code outside the package; it's great”

This can just be two sentences. “code outside the package. It’s great”

Reviewer 12: "Swift 5.9 also adds a new package access level"

Probably not necessary to mention Swift 5.9 again.

Could possibly start with “A new package access level”

Swift 5.9 supports bidirectional interoperability with C++ and Objective-C++. You can now use a subset of C++ APIs in Swift and Swift APIs from C++. For information on enabling C++ interoperability and the supported language subset, please refer to the [documentation](/documentation/cxx-interop/).

C++ interoperability is an actively evolving feature of Swift and developed by the focused [C++ Interoperability Workgroup](/cxx-interop-workgroup/). Certain aspects of the design and functionality might change in future releases of Swift, as the Swift community gathers feedback from real world adoption in mixed Swift and C++ codebases.

Join the Swift community and the C++ Interoperability Workgroup to provide feedback and help shape the evolution of the feature.

### Server

The server workgroup has been busy driving the adoption of Swift concurrency in the server ecosystem, including making use of new 5.9 features such as [custom actor executors](https://github.com/apple/swift-evolution/blob/main/proposals/0392-custom-actor-executors.md). The group recently published their [annual update for 2023](https://www.swift.org/blog/sswg-update-2023/) which goes into details of this work.

> 

Reviewer 9: ```suggestion
### Website

Along side the 5.9 release, the website workgroup has revamped the content on some of the key pages of [swift.org](http://swift.org/), including a richer home page, new detailed getting started guides, and clearer download and install instructions.

```

### Windows Platform

The Windows installation received new functionality to inject the module maps via a VFS overlay dynamically rather than needing to modify the Visual Studio installation. This change breaks the dependency order for installation; it is now possible to install either Visual Studio or the Swift toolchain first. This also eliminates the need to repair the installation after an upgrade to Visual Studio. As part of the continued refinement to the installer, initial work towards versioning the toolchain installation lays the foundation for the ability to have multiple parallel toolchain versions installed side-by-side on Windows.

> Alongside of this improvement of handling of Visual Studio, the Swift toolchain added new flags (`-windows-sdk-root`, `-windows-sdk-version`, `-visualc-tools-root`, `-visualc-tools-version`) to help control the Windows SDK and Visual C++ tools that it builds against. The Windows SDK (`WinSDK`) module also saw improvements in the coverage of the SDK enabling access to a wider set of system APIs. The Visual C++ (`vcruntime`) module was greatly restructured to support C++ interoperability.

Reviewer 5: improvement of handling -> improved handling
SDK enabling -> SDK, enabling

> Structured Concurrency is now be significantly more stable on Windows. Backtraces are still a work in progress, as the Windows ABI currently prohibits the necessary optimizations required for the proper implementation of the feature.

Reviewer 5: is now be significantly -> is now significantly

if let (int, double, string) = all(optionalInt, optionalDouble, optionalString) {
    print(int, double, string)
}
else {
    print("got a nil")
}
```

<details markdown="1" class="download">
  <summary>Implementation of <code>all</code></summary>

```swift
/// Takes any number of `Optional` arguments and returns `nil` if any of them
/// are `nil`, or a tuple of their values if all of them are non-`nil`.
func all<each Wrapped>(_ optional: repeat (each Wrapped)?) -> (repeat each Wrapped)? {
  /// Accumulates the results from the `checkNilness(of:)` helper function.
  var anyAreNil = false

  // Check if any of the arguments in `each optional` is nil by running this
  // `checkNilness(of:)` function on each argument.
  func checkNilness<OneWrapped>(of oneOptional: OneWrapped?) {
      if oneOptional == nil {
          anyAreNil = true
      }
  }
  repeat checkNilness(of: each optional)

  // Did `checkNilness(of:)` find any nils? If so, return nil early.
  if anyAreNil {
      return nil
  }

  // Otherwise, we can safely force-unwrap all of the arguments and form a tuple
  // from the unwrapped values.
  return (repeat (each optional)!)
}
```

</details>

> What's more, these features are all usable today in Swift 5.9, but they are not finished growing. Each one is too large to fully implement in a single release; we're excited about what you can already do with them in Swift 5.9, but we're even more excited to have laid the foundations for better things to come.

Reviewer 12: Maybe instead of "they are not finished growing”

“they are not finished evolving”

Growing is a little awkward, and evolving ties directly into the Swift evolution process.

> In addition to these major features, Swift 5.9 also includes smaller refinements to the language. The one developers will probably use most often is [`if` and `switch` expressions](https://github.com/apple/swift-evolution/blob/main/proposals/0380-if-switch-expressions.md), which lets you assign the result of a single-expression `if` or `switch` statement to a variable:

Reviewer 12: Maybe “includes many smaller refinements"

```swift
statusBar.text = if !hasConnection { "Disconnected" }
                 else if let error = lastError { error.localizedDescription }
                 else { "Ready" }
```

Swift 5.9 also adds a [new `package` access level](https://github.com/apple/swift-evolution/blob/main/proposals/0386-package-access-modifier.md) that lets other modules in the same package access APIs, but hides them from code outside the package; it's great for splitting up large modules into several smaller ones without exposing internals to the package's clients. Developers using Swift Concurrency may appreciate the [more convenient `DiscardingTaskGroup` types](https://github.com/apple/swift-evolution/blob/main/proposals/0381-task-group-discard-results.md) for task groups that don't generate results and the advanced [custom actor executors](https://github.com/apple/swift-evolution/blob/main/proposals/0392-custom-actor-executors.md) feature for controlling the exact environment that an actor is run in.

> A full list of Swift 5.9's Evolution proposals can be found at the end of this post.

Reviewer 12: The possessive is not necessary: "Swift 5.9 Evolution proposals"

## Developer Experience

### Crash Handling

> On Linux, the Swift runtime will now catch program crashes and Swift runtime errors and display a backtrace on the program's output. The backtracer is out-of-process and includes support for `async` functions.

Reviewer 12: Smart quote: program’s

> This feature is also available on macOS, although there it is disabled by default (it can be enabled by setting `SWIFT_BACKTRACE=enable=yes`) and you need to ensure that your program is signed with the `com.apple.security.get-task-allow` entitlement.

Reviewer 12: This is a bit awkward maybe instead:

“This feature is also available on macOS, but is disabled by default. To enable it, set `SWIFT_BACKTRACE=enable=yes` and sign your program with the `com.apple.security.get-task-allow` entitlement.”

Reviewer 12: Question: Is `SWIFT_BACKTRACE=enable=yes` an Xcode build setting? A compiler command line setting?

Reviewer 5: The latter, an environment variable it detects

### Debugging

Swift 5.9 introduces new features to [LLDB](https://lldb.llvm.org/) and the Swift compiler aimed at making Swift debugging faster and more reliable.

The `p` and `po` commands will now bypass the Swift compiler when evaluating simple expressions, which will make printing local and member variables as fast as using the `frame variable` or `v` command.

Swift expressions can now refer to generic type parameters. This allows setting a conditional breakpoint in a generic function that only triggers when a type parameter is instantiated with a specific concrete type.

Finally, the debug info produced by the Swift compiler is now more precise when it comes to scoping of local variables.

For a more detailed discussion and examples of how to use these new features, stay tuned for a dedicated blog about debugging in Swift 5.9 coming soon!

## Ecosystem

### Swift Package Manager

Following are some highlights from the changes introduced to the [Swift Package Manager](https://github.com/apple/swift-package-manager) in Swift 5.9:

> - SwiftPM packages can now use `package` as a new access modifier, allowing accessing symbols in another target / module within the same package without making it public.

Reviewer 9: ```suggestion
- Packages can make use of the new `package` access modifier, allowing access of symbols in another target / module within the same package without making it public. SwiftPM automatically sets the new compiler configuration to ensure this feature works out-of-the-box for packages. 
```

- The `CompilerPluginSupport` module enables defining macro targets. Macro targets allow authoring and distribution of custom Swift macros such as [expression macros](https://github.com/apple/swift-evolution/blob/main/proposals/0382-expression-macros.md).

- The new `.embedInCode` resource rule allows embedding the contents of the resource into the executable by generating a byte array, e.g.

  ```
  struct PackageResources {
    static let my_resource: [UInt8] = [104,101,108,108,111,32,119,111,114,108,100,10]
  }
  ```

> - The `allowNetworkConnections(scope:reason:)` setting gives a command plugin permissions to access the network. Permissions can be scoped to Unix domain sockets in general or specifically for Docker, as well as local or remote IP connections which can be limited by port. For non-interactive use cases, there is also a `--allow-network-connections` commandline flag to allow network connections for a particular scope.

Reviewer 12: “commandline flag” should be “command-line flag"

See the [Swift Package Manager changelog](https://github.com/apple/swift-package-manager/blob/main/CHANGELOG.md#swift-59) for the complete list of changes.

### Swift Syntax

[swift-syntax](https://github.com/apple/swift-syntax) has become an essential tool to create macros in Swift 5.9. In addition to the introduction of the modules that allow the creation of macros, swift-syntax has received major quality improvements:

- The documentation of swift-syntax has been greatly expanded and can be viewed at [swiftpackageindex.com](https://swiftpackageindex.com/apple/swift-syntax).

- The names of all syntax nodes and their children have been audited to be more consistent and accurately reflect the Swift language.

- The error messages produced by the new SwiftParser have been greatly improved and it now produces better errors than the C++ parser in almost all cases.

- As part of the Google Summer of Code project, incremental parsing has been introduced to swift-syntax, allowing e.g. an editor to only reparse those parts of a syntax tree that have changed.

> Over the last year, swift-syntax has been a huge success as an open source project. Since the release of Swift 5.8, more than 30 distinct open source contributors have contributed to the package accounting for more than 30% of the commits. And community tools like [swift-ast-explorer.com](https://swift-ast-explorer.com) are an invaluable tool to explore and understand the SwiftSyntax tree. Thanks to everyone who contributed!

Reviewer 12: “And” is an awkward way to begin a sentence.

Maybe instead of “And community tools” use “In addition, community tools"

---
layout: post
published: true
date: 2023-08-29 15:00:00

> title: Swift 5.9 Released!

Reviewer 13: How do others feel about dropping the exclamation mark? It matches precedent, but it's always felt a little overboard to me.

author: [alexandersandberg]
---

> Swift 5.9 is now officially released! 🎉 Swift 5.9 is the foundation for many long-awaited features that give library authors more expressive power, including a new macro system and parameter packs, improve the debugging experience through enhanced crash logs and new features in LLDB’s expression evaluator, Swift Package Manager support for a new `package` access modifier and macro targets, bi-directional interoperability with C++, Windows platform improvements, and more.

Reviewer 13: Can we punch this up and tighten it given its importance as the first paragraph? How about:

"
Swift 5.9 is now available! 🎉

This is a major new release that **adds an expressive macro system to the language** and introduces **support for mixing Swift and C++ types in the same codebase**.

It also introduces parameter packs, an impressed expression evaluator while debugging, enhanced crash logs, Windows platform improvements, and more.

Read on for a deep dive into changes to the language, standard library, tooling, platform support, and more.
"

Reviewer 7: How does it look now?

Reviewer 13: Looks great, thank you!

Thank you to everyone in the Swift community who made this release possible. Your Swift Forums discussions, bug reports, pull requests, educational content, and other contributions are always appreciated!

<!-- TODO: Link to Swift 5.9 showcase resource -->

> [The Swift Programming Language](https://docs.swift.org/swift-book/documentation/the-swift-programming-language/) book has been updated for Swift 5.9 and is now published with DocC. This is the official Swift guide and a great entry point for those new to Swift. The Swift community also maintains a number of [translations](/documentation/#translations).

Reviewer 13: Can we move this to the bottom?

## Language and Standard Library

> Swift 5.9 tackles three long-desired features that allow you to express things you couldn’t easily do before.

Reviewer 13: Instead of framing this as difficult to do before, can we frame macros and parameter packs as continuing Swift's goal of being ergonomic and easy to use, and putting `consume` last as a feature that expands for performance-critical code?

> Swift now supports [macros](https://github.com/apple/swift-evolution/blob/main/visions/macros.md), which allow developers to reduce repetitive boilerplate and create new language features that can be distributed as a Swift package. Using a macro is easy and natural; macros can either be called with the function-like freestanding `#macroName` syntax or attached to declarations with an `@MacroName` attribute, so they work just like built-in language features but can’t be mistaken for normal code. Implementing a macro uses a powerful and flexible approach: Macros are implemented by writing Swift functions that use the SwiftSyntax library to generate code to insert into the source file. Macros make it easy for your library’s users to adopt powerful capabilities that adapt to the code they’re used in, like [the new `Observation` module](https://github.com/apple/swift-evolution/blob/main/proposals/0395-observability.md), which allows Swift classes to automatically notify other code when their properties change.

Reviewer 13: Let's add a simple example of macros. It's a big, new language feature, and someone skimming this post should be able to quickly see what they look like.

Reviewer 13: Is "create new language features" the way we want to frame this? /cc @DougGregor

Reviewer 14: I think "create new language features" is a stronger claim than we want to make. They're about eliminating boilerplate and improving the expressivity of the language. What about:

```suggestion
Swift now supports [macros](https://github.com/apple/swift-evolution/blob/main/visions/macros.md), which allow developers to reduce repetitive boilerplate and create more expressive libraries that can be distributed as a Swift package. Using a macro is easy and natural; macros can either be expanded with the function-like freestanding `#macroName` syntax or attached to declarations with an `@MacroName` attribute, so they work just like built-in language features but can’t be mistaken for normal code. Implementing a macro uses a powerful and flexible approach: Macros are implemented by writing Swift functions that use the SwiftSyntax library to generate code to insert into the source file. Macros make it easy for your library’s users to adopt powerful capabilities that adapt to the code they’re used in, like [the new `Observation` module](https://github.com/apple/swift-evolution/blob/main/proposals/0395-observability.md), which allows Swift classes to automatically notify other code when their properties change.
```

Reviewer 13: Let's still add an example of using macros. It's a big new feature and it'll be helpful to see what using a macro looks like while skimming this post.

Author: Marking this as resolved in https://github.com/apple/swift-org-website/pull/372/commits/d8941008e3184f5a1966bcba41fb0d63703a3c08. Feel free to re-open if necessary.

Swift also has new features to help developers fine-tune memory management behavior in performance-critical code. The [`consume` operator](https://github.com/apple/swift-evolution/blob/main/proposals/0366-move-function.md) tells Swift to deinitialize a variable and transfer its contents without copying it. The [`consuming` and `borrowing` parameter modifiers](https://github.com/apple/swift-evolution/blob/main/proposals/0377-parameter-ownership-modifiers.md) provide hints that Swift can use to eliminate unnecessary copying and reference-counting operations when passing a parameter. Finally, [noncopyable structs and enums](https://github.com/apple/swift-evolution/blob/main/proposals/0390-noncopyable-structs-and-enums.md) allow you to create types which, like a class, represent a specific resource that can’t be meaningfully copied when it’s assigned, but like a struct or enum, do not need to be reference-counted because only one storage location can own the instance at a time.

> The last of these major features is [parameter packs](https://github.com/apple/swift-evolution/blob/main/proposals/0393-parameter-packs.md). These let you write generic types and functions which work over not just any type, but any _number_ of types. For example, this `all(_:)` function lets you check whether any number of `Optional` values are `nil`, even if the optionals wrap different types:

Reviewer 13: Parameter packs feels like a bigger feature for most Swift developers than non-copyable types. Let's move it up.

> ```swift

Reviewer 13: Is there a before and after for what parameter packs due to APIs? That would be more impactful than showing the implementation.

Reviewer 7: Yep, I'll tweak the example to show how the API is simplified with parameter packs

Reviewer 7: Updated. Do you also think the implementation should be removed?

Reviewer 13: Yeah, I think the implementation of all() is too much detail for this post, but the new before/after example looks good.

Reviewer 13: I recommend simplifying the "after" example too by dropping the declarations of optionalInt, optionalDouble, etc. I think those can be implied.

if let (int, double, string) = all(optionalInt, optionalDouble, optionalString) {
    print(int, double, string)
}
else {
    print("got a nil")
}
```

<details markdown="1" class="download">
  <summary>Implementation of <code>all</code></summary>

```swift
/// Takes any number of `Optional` arguments and returns `nil` if any of them
/// are `nil`, or a tuple of their values if all of them are non-`nil`.
func all<each Wrapped>(_ optional: repeat (each Wrapped)?) -> (repeat each Wrapped)? {
  /// Accumulates the results from the `checkNilness(of:)` helper function.
  var anyAreNil = false

  // Check if any of the arguments in `each optional` is nil by running this
  // `checkNilness(of:)` function on each argument.
  func checkNilness<OneWrapped>(of oneOptional: OneWrapped?) {
      if oneOptional == nil {
          anyAreNil = true
      }
  }
  repeat checkNilness(of: each optional)

  // Did `checkNilness(of:)` find any nils? If so, return nil early.
  if anyAreNil {
      return nil
  }

  // Otherwise, we can safely force-unwrap all of the arguments and form a tuple
  // from the unwrapped values.
  return (repeat (each optional)!)
}
```

</details>

> What’s more, these features are all usable today in Swift 5.9, but they are not finished evolving. Each one is too large to fully implement in a single release; we’re excited about what you can already do with them in Swift 5.9, but we’re even more excited to have laid the foundations for better things to come.

Reviewer 13: I don't think this sentence is necessary. It doesn't explain how these language features will change and only makes me worry they're not ready to be used.

Reviewer 7: I refined this to make it focus more on the foundation provided in Swift 5.9, and make it clear that the future is additive as opposed to these features being "unfinished" in Swift 5.9. What do you think?

Reviewer 7: I'm open to removing this, but I attempted to keep it in because the Language Steering Group liked this message when we reviewed the post this week. However, I can also see how this message fits more naturally into a "next year in the Swift project" blog post similar to the one from last year, as opposed to the 5.9 release blog post.

Author: I like the revised sentence in https://github.com/apple/swift-org-website/pull/372/commits/4f13f198b98b2d011305b9b32e29405d4130ecfe, and think it makes sense to keep in the blog post. 👍 Marking this as resolved, but @cthielen feel free to review and re-open if necessary.

> In addition to these major features, Swift 5.9 also includes many smaller refinements to the language. The one developers will probably use most often is [`if` and `switch` expressions](https://github.com/apple/swift-evolution/blob/main/proposals/0380-if-switch-expressions.md), which lets you assign the result of a single-expression `if` or `switch` statement to a variable:

Reviewer 13: You can tighten this up:

"Swift 5.9 also includes smaller, quality-of-life changes, like the ability to use if and switch expressions to assign a result: <code sample>"

Developers using Swift Concurrency may appreciate the [more convenient `DiscardingTaskGroup` types](https://github.com/apple/swift-evolution/blob/main/proposals/0381-task-group-discard-results.md) for task groups that don’t generate results and the advanced [custom actor executors](https://github.com/apple/swift-evolution/blob/main/proposals/0392-custom-actor-executors.md) feature for controlling the exact environment that an actor is run in.

A full list of Swift 5.9 Evolution proposals can be found at the end of this post.

## Developer Experience

### Crash Handling

On Linux, the Swift runtime will now catch program crashes and Swift runtime errors and display a backtrace on the program’s output. The backtracer is out-of-process and includes support for `async` functions.

This feature is also available on macOS, but is disabled by default. To enable it, set `SWIFT_BACKTRACE=enable=yes` and sign your program with the `com.apple.security.get-task-allow` entitlement.

### Debugging

Swift 5.9 introduces new features to [LLDB](https://lldb.llvm.org/) and the Swift compiler aimed at making Swift debugging faster and more reliable.

> The `p` and `po` commands will now bypass the Swift compiler when evaluating simple expressions, which will make printing local and member variables as fast as using the `frame variable` or `v` command.

Reviewer 13: The "what" and "how" are flipped here. Better to say, `p` and `po` print local and members variables much faster now by bypassing ...

Swift expressions can now refer to generic type parameters. This allows setting a conditional breakpoint in a generic function that only triggers when a type parameter is instantiated with a specific concrete type.

> Finally, the debug info produced by the Swift compiler is now more precise when it comes to scoping of local variables.

Reviewer 13: More precise how, or why? Also, drop "Finally".

Reviewer 7: Good question. @adrian-prantl can you please describe the concrete improvement for users of LLDB?

> For a more detailed discussion and examples of how to use these new features, stay tuned for a dedicated blog about debugging in Swift 5.9 coming soon!

Reviewer 13: I don't think this adds anything and should be dropped. Folks will see the blog post when it comes out.

## Ecosystem

### Swift Package Manager

The following are some highlights from the changes introduced to the [Swift Package Manager](https://github.com/apple/swift-package-manager) in Swift 5.9:

- Packages can make use of the new `package` access modifier, allowing access of symbols in another target / module within the same package without making them public. SwiftPM automatically sets the new compiler configuration to ensure this feature works out-of-the-box for packages.

- The `CompilerPluginSupport` module enables defining macro targets. Macro targets allow authoring and distribution of custom Swift macros such as [expression macros](https://github.com/apple/swift-evolution/blob/main/proposals/0382-expression-macros.md).

> - The new `.embedInCode` resource rule allows embedding the contents of the resource into the executable by generating a byte array, e.g.

Reviewer 13: I'm not familiar with this. Is the code sample below the result of using `.embedInCode`?

Reviewer 7: I found the PR that implemented this and it looks like you use `.embedInCode` in a package manifest like this

```swift
  let package = Package(
    name: "MyPackage",
    targets: [
      .executableTarget(name: "MyTarget", resources: [.embedInCode("path/to/resource.txt")]),
    ]
  )
```

And the example below is the code that's generated for you. I think the example doesn't add much, might be best to drop it.

Reviewer 9: cc @neonichu to say

Reviewer 13: I recommend either dropping the example or swapping it out for the one you have above. Better to show the "call site" than the result.

Reviewer 7: I dropped the example already 👍

  ```
  struct PackageResources {
    static let my_resource: [UInt8] = [104,101,108,108,111,32,119,111,114,108,100,10]
  }
  ```

> - The `allowNetworkConnections(scope:reason:)` setting gives a command plugin permissions to access the network. Permissions can be scoped to Unix domain sockets in general or specifically for Docker, as well as local or remote IP connections which can be limited by port. For non-interactive use cases, there is also a `--allow-network-connections` command-line flag to allow network connections for a particular scope.

Reviewer 13: Can we say why this is useful? e.g. This new command plugin permission lets you ...

> - SwiftPM can now publish to a registry following the publishing spec as defined in [SE-0391](https://github.com/apple/swift-evolution/blob/main/proposals/0391-package-registry-publish.md). SwiftPM also gains support for signed packages. Trust-on-first-use (TOFU) validation checks which previously only included fingerprints (e.g., checksums) has been extended to include signing identities, and it is enforced for source archives as well as package manifests.

Reviewer 13: Should this be two list items, registry and signed packages? If not, let's put them together, e.g. "SwiftPM can now publish signed packages to a registry as defined in SE …"

Reviewer 9: they are related, signed packages are a feature of regsitry

Reviewer 7: I clarified the text to say that a registry may require signed packages (based on the text in SE-0391), let me know what you think

> - A new `swift experimental-sdk` experimental command is now available for managing Swift SDK bundles that follow the format described in [SE-0387](https://github.com/apple/swift-evolution/blob/main/proposals/0387-cross-compilation-destinations.md).

Reviewer 13: Is the command itself experimental, or is it a stable command that enables an experimental SDK? If the latter, you should drop the word "experimental" before command.

Reviewer 13: Since the SE proposals are at the bottom of the post, I don't think you should reference them like "… as described in SE…". You could specify the link on the text "Swift SDK bundles" instead.

Reviewer 7: I actually don't see any mention of `swift experimental-sdk` in SE-0387, but I see many mentions of "Swift SDK" without the experimental. Perhaps this bullet (which looks like it was copied from the SwiftPM changelog) is a remnant of before the proposal was accepted. I'll update it according to my understanding of SE-0387.

Reviewer 9: the feature is still considered experimental, which is why the command is named that way - to signal the users it may slightly evolve still.

Reviewer 9: we can consider not mentioning the command by name here to avoid confusion

Reviewer 7: Is this command documented anywhere? I find it confusing that this bullet links to SE-0387 if the implementation does not match the proposal.

Reviewer 15: Shouldn't we have dropped the `experimental` after evolution acceptance?

Reviewer 6: > Is this command documented anywhere? I find it confusing that this bullet links to SE-0387 if the implementation does not match the proposal.

It's documented in the same way as the other subcommands, with ArgumentParser help strings.

> Shouldn't we have dropped the experimental after evolution acceptance?

Maybe @tomerd can clarify on this one. AFAIR, the intention was to keep the experimental prefix until we feel that the implementation and generated Swift SDKs are solid and work end-to-end in all intended cases, including static linking, and that we can guarantee that specifically linker behavior when cross-compiling can be identical to native builds.

Reviewer 9: even though the proposal was accepted and the implementation matches it to a large degree, the feature is still considered experimental due to remaining decisions around shipping a linker in the toolchain to better support the feature.

in the context of the blog post, if we feel mentioning features that are still experimental is confusing, we can drop the mention all together. otherwise, open to how to better describe it

Reviewer 13: I see, that is a bit confusing. It might be just 5% better to write: "An experiment, new `swift experimental-sdk` command ..."

Reviewer 9: how about:

```
- SwiftPM now supports cross compilation based on Swift  SDKs as defined in [SE-0387](https://github.com/apple/swift-evolution/blob/main/proposals/0387-cross-compilation-destinations.md). While feature is still considered experimental, we invite users to try it out and provide feedback
```

cc @hborla

Author: Marking this as resolved in https://github.com/apple/swift-org-website/pull/372/commits/902fa61a3a00330d63bef1454f569ca94dd706eb.

See the [Swift Package Manager changelog](https://github.com/apple/swift-package-manager/blob/main/CHANGELOG.md#swift-59) for the complete list of changes.

### Swift Syntax

> [swift-syntax](https://github.com/apple/swift-syntax) has become an essential tool to create macros in Swift 5.9. In addition to the introduction of the modules that allow the creation of macros, swift-syntax has received major quality improvements:

Reviewer 13: How about: "swift-syntax is an essential tool for parsing Swift code, and helps to power new language macros. This year, it received a number of major improvements:"

> - The documentation of swift-syntax has been greatly expanded and can be viewed at [swiftpackageindex.com](https://swiftpackageindex.com/apple/swift-syntax).

Reviewer 13: This should come last

> - The names of all syntax nodes and their children have been audited to be more consistent and accurately reflect the Swift language.

Reviewer 13: "Audited" isn't needed. "Syntax node names are more consistent and accurately reflect the Swift language"

> - The error messages produced by the new SwiftParser have been greatly improved and it now produces better errors than the C++ parser in almost all cases.

Reviewer 13: The compare to the C++ parser isn't necessary. Also, can we state how the error messages are better?

> - As part of the Google Summer of Code project, incremental parsing has been introduced to swift-syntax, allowing e.g. an editor to only reparse those parts of a syntax tree that have changed.

Reviewer 13: Lose the GSoC reference: "swift-syntax now supports incremental parsing, allowing an editor to only re-parse …"

Over the last year, swift-syntax has been a huge success as an open source project. Since the release of Swift 5.8, more than 30 distinct open source contributors have contributed to the package accounting for more than 30% of the commits. In addition, community tools like [swift-ast-explorer.com](https://swift-ast-explorer.com) are an invaluable tool to explore and understand the SwiftSyntax tree. Thanks to everyone who contributed!

> ### C++ Interoperability

Reviewer 13: Shouldn't this be at the top of Ecosystem? It certainly seems more impactful than SwiftPM and swift-syntax improvements.

> Swift 5.9 supports bidirectional interoperability with C++ and Objective-C++. You can now use a subset of C++ APIs in Swift and Swift APIs from C++. For information on enabling C++ interoperability and the supported language subset, please refer to the [documentation](/documentation/cxx-interop/).

Reviewer 13: Can we show a code example here?

C++ interoperability is an actively evolving feature of Swift and developed by the focused [C++ Interoperability Workgroup](/cxx-interop-workgroup/). Certain aspects of the design and functionality might change in future releases of Swift, as the Swift community gathers feedback from real world adoption in mixed Swift and C++ codebases.

Join the Swift community and the C++ Interoperability Workgroup to provide feedback and help shape the evolution of the feature.

> ### Server

Reviewer 13: What's the Swift 5.9 feature here?

Reviewer 9: discussed with @tkremenek and we thought it would be worth to mention even though not a 5.9 feature per-se

Reviewer 13: I see. How about we restate this in the framing of Swift 5.9 features then: "Custom Actor Executors and other features from Swift 5.9 are making their way into the Swift on server ecosystem by <example here>. The server workgroup also recently published their annual update for 2023, detailing efforts to increase adoption of concurrency within key libraries and more."

Reviewer 9: I like that!

The server workgroup has been busy driving the adoption of Swift concurrency in the server ecosystem, including making use of new 5.9 features such as [custom actor executors](https://github.com/apple/swift-evolution/blob/main/proposals/0392-custom-actor-executors.md). The group recently published their [annual update for 2023](/blog/sswg-update-2023/) which goes into details of this work.

> ### Website

Reviewer 13: What's the Swift 5.9 feature here?

Reviewer 9: same as the above re. server

Reviewer 7: How about we move the Website, Downloads, and Language Guide sections together into a "Next Steps" or "Get Started with Swift 5.9" section at the end of the post? I do think we should direct readers of this blog post at the new Swift.org landing page -- which is a new feature of the Website developed alongside Swift 5.9 -- because it's a much more useful tool for getting started with Swift depending on your use case.

Reviewer 9: yea that a cool idea

Author: That's a great idea. "Next Steps" might hint towards "next steps for the Swift language" rather than "next steps for those wanting to get started with Swift 5.9", so maybe naming the section "Get Started" or something similar is better?

Along with the 5.9 release, the website workgroup has revamped the content on key pages of [Swift.org](/), including a richer home page, new and detailed getting-started guides, and clearer download and install instructions.

### Windows Platform

> The Windows installation received new functionality to inject the module maps via a VFS overlay dynamically rather than needing to modify the Visual Studio installation. This change breaks the dependency order for installation; it is now possible to install either Visual Studio or the Swift toolchain first. This also eliminates the need to repair the installation after an upgrade to Visual Studio. As part of the continued refinement to the installer, initial work towards versioning the toolchain installation lays the foundation for the ability to have multiple parallel toolchain versions installed side-by-side on Windows.

Reviewer 13: This paragraph also swaps the "what" and the "how". Better to start with "The Windows installer now supports installation both before and after Visual Studio installation, and no longer requires repair after a Visual Studio upgrade. This is made possible by …"

Alongside this improved handling of Visual Studio, the Swift toolchain added new flags (`-windows-sdk-root`, `-windows-sdk-version`, `-visualc-tools-root`, `-visualc-tools-version`) to help control the Windows SDK and Visual C++ tools that it builds against. The Windows SDK (`WinSDK`) module also saw improvements in the coverage of the SDK, enabling access to a wider set of system APIs. The Visual C++ (`vcruntime`) module was greatly restructured to support C++ interoperability.

> Structured Concurrency is now significantly more stable on Windows. Backtraces are still a work in progress, as the Windows ABI currently prohibits the necessary optimizations required for the proper implementation of the feature.

Reviewer 13: Since backtraces aren't ready, we don't need to mention them in this post.

Reviewer 13: Can we state how it is more stable, or give numbers?

Reviewer 7: @compnerd do you have any concrete examples of the stability improvement here?

Reviewer 16: A concrete example would be something like a simple for loop that called `await` would previously stack overflow and no longer will (e.g. awaiting ~350 values from `AsyncStream`). This is because we opted to break the OS ABI and perform an invalid operation to do the tailcall on the resumption function and thus lost stack traces (the stack is no longer valid as far as Windows is concerned).

Reviewer 16: ```swift
@main
struct S {
  public static func main() async {
    let stream: AsyncStream<Int> = AsyncStream<Int> { continuation in
      Task.detached {
        var sum: Int = 0
        for i in 0 ..< 100_000 {
          let value = i
          sum += value
          continuation.yield(value)
        }
        continuation.yield(-sum)
        continuation.finish()
      }

      var sum: Int = 0
      for await value in stream {
        sum += value
      }
    }
  }
}
```

Is a more concrete example that would have previously failed to execute and should now be fine.

Reviewer 7: Thank you!

- A new `swift sdk` command is now available for installing and removing [Swift SDK bundles](https://github.com/apple/swift-evolution/blob/main/proposals/0387-cross-compilation-destinations.md#swift-sdk-bundles) on the local filesystem.

See the [Swift Package Manager changelog](https://github.com/apple/swift-package-manager/blob/main/CHANGELOG.md#swift-59) for the complete list of changes.

### Swift Syntax

[swift-syntax](https://github.com/apple/swift-syntax) is an essential tool for parsing Swift code, and it helps to power new macro libraries. This year, swift-syntax received a number of major improvements:

- Syntax nodes names are more consistent and accurately reflect the Swift language.

- Error recovery in the new SwiftParser is greatly improved, leading to more precise error messages for incorrect or missing syntax.

- Incremental parsing is now supported, allowing tools such as an editor to only reparse those parts of a syntax tree that have changed.

> - The documentation of swift-syntax has been greatly expanded and can be viewed at [swiftpackageindex.com](https://swiftpackageindex.com/apple/swift-syntax).

Reviewer 17: i know this is quite late, but have we thought about changing the link to point to the [SwiftSyntax docs on Swiftinit](https://swiftinit.org/docs/swift-syntax/swiftsyntax)? as SwiftSyntax has a lot of API spread out across multiple modules, the Unidoc-generated documentation has [a lot of features](https://forums.swift.org/t/introducing-unidoc-scalable-multi-package-swift-documentation/67287) for this kind of use case, that aren’t supported by DocC documentation just yet. notably, the interactive search bar shows results from *all* the SwiftSyntax modules, instead of just the current module.

Author: cc: @ahoppen @bnbarham

---
layout: post
published: true
date: 2023-08-29 15:00:00
title: Swift 5.9 Released
author: [alexandersandberg, hborla]
---

Swift 5.9 is now available! 🎉

This is a major new release that **adds an expressive macro system to the language** and introduces **support for integrating Swift into C++ codebases** through bidirectional interoperability.

It also introduces parameter packs, an improved expression evaluator while debugging, enhanced crash handling, Windows platform improvements, and more.

Read on for a deep dive into changes to the language, standard library, tooling, platform support, and next steps for getting started with Swift 5.9.

Thank you to everyone in the Swift community who made this release possible. Your Swift Forums discussions, bug reports, pull requests, educational content, and other contributions are always appreciated!

<!-- TODO: Link to Swift 5.9 showcase resource -->

## Language and Standard Library

Swift's fundamental goal is to encourage code that is clear and concise, while remaining safe and efficient.

> This release adds three long-desired features that further that goal: a new macro system for more expressive libraries, parameter packs to make overloaded APIs natural to use, and new ownership features to offer more performance control in low-level code.

Reviewer 5: more performance control in -> more control over performance  in

A full list of Swift 5.9 Evolution proposals can be found at the end of this post.

## Developer Experience

### Crash Handling

On Linux, the Swift runtime will now catch program crashes and Swift runtime errors and display a backtrace on the program’s output. The backtracer is out-of-process and includes support for `async` functions.

This feature is also available on macOS but is disabled by default. To enable it, set `SWIFT_BACKTRACE=enable=yes` and sign your program with the `com.apple.security.get-task-allow` entitlement.

### Debugging

Swift 5.9 introduces new features to [LLDB](https://lldb.llvm.org/) and the Swift compiler aimed at making Swift debugging faster and more reliable.

The `p` and `po` commands now print local variables and properties as fast as the `frame variable` or `v` commands by bypassing the Swift compiler when evaluating simple expressions.

Swift expressions can now refer to generic type parameters. This allows setting a conditional breakpoint in a generic function that only triggers when a type parameter is instantiated with a specific concrete type.

> The debug info produced by the Swift compiler is now more precisely scoping local variables, which makes it less likely to see variables backed by uninitialized memory in the debugger.

Reviewer 5: is now more precisely scoping local variables, which makes -> now more precisely scopes local variables, making
likely to see -> likely that you see

```

you can call it directly from your Swift code:

```swift
import PromptResponder

let codeLines = generatePromptResponse("Write Swift code that prints hello world")
  .map(String.init)
  .filter { !$0.isEmpty }

for line in codeLines {
  print(line)
}
```

C++ interoperability is actively evolving, with some aspects subject to change in future releases as the community gathers feedback from real world adoption in mixed Swift and C++ codebases.

For information on enabling C++ interoperability and the supported language subset, please refer to the [documentation](/documentation/cxx-interop/).

Special thanks to the [C++ Interoperability Workgroup](/cxx-interop-workgroup/) for their focus and dedication in supporting this feature.

### Swift Package Manager

The [Swift Package Manager](https://github.com/apple/swift-package-manager) has a number of improvements in Swift 5.9:

- Packages can use the new `package` access modifier, allowing access of symbols in another target / module within the same package without making them public. SwiftPM automatically sets the new compiler configuration to ensure this feature works out-of-the-box for packages.

> - The `CompilerPluginSupport` module enables defining macro targets. Macro targets allow authoring and distributing custom Swift macros as API in a library.

Reviewer 5: API - > APIs

- The new `.embedInCode` resource rule allows embedding the contents of a resource into an executable by generating a byte array.

- The `allowNetworkConnections(scope:reason:)` setting gives a command plugin permissions to access the network. Permissions can be scoped to Unix domain sockets as well as local or remote IP connections, with an option to limit by port. For non-interactive use cases, the `--allow-network-connections` command-line flag allows network connections for a particular scope.

- SwiftPM can now publish to a registry following the specification defined in [SE-0391](https://github.com/apple/swift-evolution/blob/main/proposals/0391-package-registry-publish.md), as well as support signed packages, which may be required by a registry. Trust-on-first-use (TOFU) validation checks can now use signing identities in addition to fingerprints, and are enforced for source archives as well as package manifests.

> - SwiftPM now supports cross compilation based on [Swift SDK bundles](https://github.com/apple/swift-evolution/blob/main/proposals/0387-cross-compilation-destinations.md#swift-sdk-bundles). While the feature is still considered experimental, we invite users to try it out and provide feedback.

Reviewer 5: Swift SDK bundles -> the new Swift SDK bundle format

See the [Swift Package Manager changelog](https://github.com/apple/swift-package-manager/blob/main/CHANGELOG.md#swift-59) for the complete list of changes.

### Swift Syntax

[swift-syntax](https://github.com/apple/swift-syntax) is an essential tool for parsing Swift code and helps power the new macro system. This year, swift-syntax received a number of improvements:

> - Syntax nodes names are more consistent and accurately reflect the Swift language.

Reviewer 5: Syntax nodes names -> Names of Syntax nodes

Alternatively -> Syntax nodes' names

The current formulation is wrong.

Thank you to everyone in the Swift community who made this release possible. Your Swift Forums discussions, bug reports, pull requests, educational content, and other contributions are always appreciated!

<!-- TODO: Link to Swift 5.9 showcase resource -->

[The Swift Programming Language](https://docs.swift.org/swift-book/documentation/the-swift-programming-language/) book has been updated for Swift 5.9 and is now published with DocC. This is the official Swift guide and a great entry point for those new to Swift. The Swift community also maintains a number of [translations](/documentation/#translations).

If you’re new to Swift, [The Swift Programming Language](https://docs.swift.org/swift-book/) is the official Swift guide and has been updated for version 5.9. The Swift community maintains a number of [translations](https://www.swift.org/documentation/#translations).

## Language and Standard Library

<!-- TODO: Add section content -->

You can find the complete list of Swift Evolution proposals that were implemented in Swift 5.9 in the [Swift Evolution Appendix](#swift-evolution-appendix) below.

## Developer Experience

<!-- TODO: Add more sections with content? -->

> ### Debugging

Author: @adrian-prantl @augusto2112 @kastiglione I quickly wrote a short introduction to your blog post before linking to it. Feel free to adjust. 🙂

(I'm assuming your blog post will be published before this announcement. If not, we'll have to revisit this.)

New features have been introduced to [LLDB](https://lldb.llvm.org/ "LLDB project home page") and the Swift compiler to improve the debugging experience, including faster `p` and `po` and use of generic type parameters in expressions.

Learn more in this dedicated blog post: [What’s new in Swift debugging on the 5.9 branch?](#). <!-- TODO: add link to blog post once it's published -->

## Ecosystem

<!-- TODO: Add more sections with content? -->

> ### Swift Package Manager

Author: @tomerd Let me know if there's anything I can improve for this SwiftPM section. 🙂

Reviewer 9: cc @neonichu @yim-lee @MaxDesiatov

Reviewer 9: https://github.com/apple/swift-org-website/pull/374

- New `swift experimental-sdk` experimental command is now available for managing Swift SDK bundles that follow the format described in [SE-0387](https://github.com/apple/swift-evolution/blob/main/proposals/0387-cross-compilation-destinations.md).

- SwiftPM can now publish to a registry following the publishing spec as defined in [SE-0391](https://github.com/apple/swift-evolution/blob/main/proposals/0391-package-registry-publish.md). SwiftPM also gains support for signed packages. Trust-on-first-use (TOFU) check which includes only fingerprints (e.g., checksums) previously has been extended to include signing identities, and it is enforced for source archives as well as package manifests.

- Add a new `CompilerPluginSupport` module which contains the definition for macro targets. Macro targets allow authoring and distribution of custom Swift macros such as [expression macros](https://github.com/apple/swift-evolution/blob/main/proposals/0382-expression-macros.md).

See the [Swift Package Manager changelog](https://github.com/apple/swift-package-manager/blob/main/CHANGELOG.md#swift-59) for the complete list of changes.

> ### SwiftSyntax

Author: @ahoppen @bnbarham I have added the content from you. I changed some **swift-syntax** to **SwiftSyntax** to keep it consistent, but feel free to update if necessary. 🙂

Reviewer 18: Thanks! SwiftSyntax is fine with me :)

Reviewer 19: I prefer to keep swift-syntax where I wrote swift-syntax. SwiftSyntax is just the SwiftSyntax module while swift-syntax is the entire package (which contains the SwiftSyntax module but also SwiftParser, SwiftSyntaxBuilder, SwiftSytnaxMacros etc.). And this section covers the entire package not just the SwiftSyntax module of the package.

Author: @ahoppen I see—my bad! I will change it back. Then it probably also makes sense to title the section "swift-syntax", and not "SwiftSyntax"?

Reviewer 18: You could say the same thing about the other similar cases @ahoppen. Honestly I'm not sure it really matters 😅 We could go with "Swift Syntax" as well just to add to the options.

Reviewer 18: I spoke with Alex this morning, the confusing part with swift-syntax is that "Swift Syntax" could refer to either the syntax of Swift or the swift-syntax package. The uppercase "Syntax" generally disambiguates this (being the package), but it is fairly subtle. So let's stay with "swift-syntax" in the body text. Unfortunately that looks a bit weird in the title, so "Swift Syntax" will have to do there, that seems fine since the very first sentence disambiguates what it's about.

Author: Perfect, thanks for clearing that up! I have updated the section: https://github.com/apple/swift-org-website/pull/372/commits/7518f22dae2c2ab8320d0184ddb0e9fbedc36d13

Will resolve this conversation, but let me know if something needs to be adjusted. 🙂

Swift 5.9 introduces new features to [LLDB](https://lldb.llvm.org/) and the Swift compiler aimed at making Swift debugging faster and more reliable.

The `p` and `po` commands now print local variables and properties as fast as the `frame variable` or `v` commands by bypassing the Swift compiler when evaluating simple expressions.

Swift expressions can now refer to generic type parameters. This allows setting a conditional breakpoint in a generic function that only triggers when a type parameter is instantiated with a specific concrete type.

The debug info produced by the Swift compiler is now more precisely scoping local variables, which makes it less likely to see variables backed by uninitialized memory in the debugger.

## Ecosystem

### C++ Interoperability

Swift 5.9 supports bidirectional interoperability with C++ and Objective-C++. You can now use a subset of C++ APIs in Swift and Swift APIs from C++:

> ```cpp

Reviewer 13: I like the sample! But how about being even more direct, e.g.:

"
Swift 5.9 supports bidirectional interoperability with C++ and Objective-C++ for certain kinds of APIs.

Now, if you had a C++ function like:
```
std::vector<std::string> generatePromptResponse(std::string prompt);
```

you could call it directly from your Swift code:

```
let codeLines = generatePromptResponse("Write Swift code that prints hello world")
  .map(String.init)
  .filter { !$0.isEmpty }
```

Author: Marking this as resolved in https://github.com/apple/swift-org-website/pull/372/commits/5838920a7545404a3079278e7ca668c6cc5d811c.
